Domain/SecurityIdentityRetrievalStrategy.php

```diff
@@ -39,37 +39,54 @@ public function __construct(RoleHierarchyInterface $roleHierarchy, Authenticatio
 
     /**
      * {@inheritdoc}
+     *
+     * @return RoleSecurityIdentity[]
      */
-    public function getSecurityIdentities(TokenInterface $token)
+    public function getSecurityIdentities(?TokenInterface $token)
     {
         $sids = [];
 
         // add user security identity
-        if (!$token instanceof AnonymousToken) {
+        if (!$token instanceof AnonymousToken && null !== $token) {
             try {
                 $sids[] = UserSecurityIdentity::fromToken($token);
             } catch (\InvalidArgumentException $e) {
                 // ignore, user has no user security identity
             }
         }
 
-        // add all reachable roles
-        foreach ($this->roleHierarchy->getReachableRoleNames($token->getRoleNames()) as $role) {
-            $sids[] = new RoleSecurityIdentity($role);
+        if (null !== $token) {
+            // add all reachable roles
+            foreach ($this->roleHierarchy->getReachableRoleNames($token->getRoleNames()) as $role) {
+                $sids[] = new RoleSecurityIdentity($role);
+            }
         }
 
+        $anonymousRole = \defined('\Symfony\Component\Security\Core\Authorization\Voter\AuthenticatedVoter::PUBLIC_ACCESS') ? AuthenticatedVoter::PUBLIC_ACCESS : AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY;
         // add built-in special roles
         if ($this->authenticationTrustResolver->isFullFledged($token)) {
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_FULLY);
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_REMEMBERED);
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
         } elseif ($this->authenticationTrustResolver->isRememberMe($token)) {
             $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_REMEMBERED);
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
-        } elseif ($this->authenticationTrustResolver->isAnonymous($token)) {
-            $sids[] = new RoleSecurityIdentity(AuthenticatedVoter::IS_AUTHENTICATED_ANONYMOUSLY);
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
+        } elseif ($this->isNotAuthenticated($token)) {
+            $sids[] = new RoleSecurityIdentity($anonymousRole);
         }
 
         return $sids;
     }
+
+    /**
+     * @psalm-suppress UndefinedInterfaceMethod
+     */
+    private function isNotAuthenticated(?TokenInterface $token): bool
+    {
+        if (\defined('\Symfony\Component\Security\Core\Authorization\Voter\AuthenticatedVoter::PUBLIC_ACCESS')) {
```

Comment:
wouldnt be better to check for method exists here instead?

Reply:
With the method_exists check it will still use the old method on symfony 5.4 as the new method couldn't be added in the interface and it's just a `@method` annotation. With this check we are able to switch to the new method as soon as it is available (that constant was also added in symfony 5.4)

https://github.com/symfony/symfony/blob/143ad6a04bbd7e8e73e35afb3d4c2438886273a2/src/Symfony/Component/Security/Core/Authentication/AuthenticationTrustResolverInterface.php#L21

Reply:
Oh my bad I forgot about that.